language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/JoinClauseNodeContext.java

```diff
@@ -115,10 +129,65 @@ private boolean onSuggestBindingPattern(BallerinaCompletionContext context, Join
         int cursor = context.getCursorPositionInTree();
         TypedBindingPatternNode typedBindingPattern = node.typedBindingPattern();
-        ExpressionNode expression = node.expression();
 
-        return node.inKeyword().isMissing() && !typedBindingPattern.isMissing()
-                && cursor > typedBindingPattern.textRange().endOffset()
-                && (expression.isMissing() || cursor < expression.textRange().startOffset());
+        if (!node.inKeyword().isMissing() || typedBindingPattern.isMissing()) {
+            return false;
+        } else if (cursor > typedBindingPattern.textRange().endOffset() &&
+                cursor <= node.expression().textRange().startOffset()) {
+            /*
+             * Captures:
+             * (1) join var varName <cursor>
+             */
+            return true;
+        } else {
+            Node nodeAtCursor = context.getNodeAtCursor();
+
+            /*
+             * Captures:
+             * (1) join var varName i<cursor>
+             * (2) join var varName i<cursor> expression
+             */
+            if (nodeAtCursor.kind() == SyntaxKind.SIMPLE_NAME_REFERENCE) {
+                SimpleNameReferenceNode nameReferenceNode = (SimpleNameReferenceNode) nodeAtCursor;
+                return node.expression().textRange().startOffset() + 1 == cursor &&
+                        nameReferenceNode.textRange().startOffset() + 1 == cursor &&
+                        ItemResolverConstants.IN_KEYWORD.startsWith(nameReferenceNode.name().text());
+            }
+        }
+
+        return false;
+    }
+
+    private boolean onSuggestOnKeyword(BallerinaCompletionContext context, JoinClauseNode node) {
+        int cursor = context.getCursorPositionInTree();
+
+        if (node.typedBindingPattern().isMissing() ||
+                node.typedBindingPattern().isMissing() ||
+                node.inKeyword().isMissing() ||
+                node.expression().isMissing()) {
+            return false;
+        } else if (!node.joinOnCondition().onKeyword().isMissing()) {
+            return false;
+        } else if (node.expression().textRange().endOffset() < cursor &&
+                cursor <= node.joinOnCondition().textRange().startOffset()) {
+            // join var varName in expr <cursor>
+            return true;
+        } else {
+            Node nodeAtCursor = context.getNodeAtCursor();
+
+            /*
+             * Captures:
+             * (1) join var varName in expr o<cursor>
+             * (2) join var varName in expr o<cursor> expr
+             */
+            if (nodeAtCursor.kind() == SyntaxKind.SIMPLE_NAME_REFERENCE) {
+                SimpleNameReferenceNode nameReferenceNode = (SimpleNameReferenceNode) nodeAtCursor;
+                return node.expression().textRange().endOffset() == cursor &&
+                        nameReferenceNode.textRange().endOffset() == cursor &&
+                        ItemResolverConstants.ON.startsWith(nameReferenceNode.name().text());
```

Comment:
Same as in the previous comment

Reply:
Removed the `startsWith()` check. See above thread for more details.